### PR TITLE
Update dependant nodes for ODP

### LINKED ISSUE #1627: ODP nodes update: calculate and validate dependents nodes
when updating / deleting odp,
calculate nodes that are dependents of all variables of forestArea and forestCharacteristics .
it should be the same than updateNodesValues. but with predefined variables.

- [x] on update
- [ ] on delete

### LINKED ISSUE #1666: use web socket to update OPD visible in tables 1a and 1b on odp changes
- [x] on update
- [x] on delete

### PR DESCRIPTION
**Pull request by sorja**
# Todo:
- [x] Merge 1a and 1b node updates to avoid double update for node

Resolves #1627

https://user-images.githubusercontent.com/5508251/198965817-238c2d22-20b0-48fc-a3f1-785ddaa3ba77.mov

Resolves #1666

https://user-images.githubusercontent.com/5508251/199221045-90919871-dc32-4a8d-9a87-0f669f74c685.mov

### CODE REVIEW

**Review by rfontanarosa — APPROVED** [review]

File: src/client/store/pages/assessmentSection/slice.ts [hunk 1]
```diff
@@ -37,8 +37,19 @@ export const assessmentSectionSlice = createSlice({
...
+        const dataUpdate = TableDatas.updateDatum({
+          ...dataUpdateProps,
+          data: state.originalDataPointData,
+          tableName: 'originalDataPointValue',
+        })
+        state.originalDataPointData = { ...state.originalDataPointData, ...dataUpdate }
+      } else {
+        const dataUpdate = TableDatas.updateDatum(dataUpdateProps)
+        state.data = { ...data, ...dataUpdate }
+      }
```
Comment (sorja) [inline src/client/store/pages/assessmentSection/slice.ts]: When we update->receive an ODP value from websocket, handle it here

Although, when writing this and working with the websockets, I was wondering should we just use one state variable (eg. `data` here) and receive the data directly from backend?

Do we ever use non-odp values for odp years when 1b has ODPs enabled?

File: src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts [hunk 2]
```diff
@@ -7,7 +7,10 @@ import { RowRepository } from '@server/repository/assessment/row'
 
 import { calculateNode } from './calculateNode'
 
-export const calculateDependantNodes = async (props: Props, client: BaseProtocol): Promise<NodeUpdates> => {
+export const calculateDependantNodes = async (
+  props: Omit<Props, 'value'>,
```
Comment (sorja) [inline src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts:11]: Omit property `value` as it was not used and we did not have it in updateOriginalDataPoint

File: src/server/controller/cycleData/updateOriginalDataPoint.ts [hunk 3]
```diff
@@ -16,11 +185,34 @@ export const updateOriginalDataPoint = async (
...
+
+    await updateForestCharacteristicsDependants({
+      countryIso,
+      assessment,
+      cycle,
+      colName,
+      user,
+      updatedOriginalDataPoint,
+      client: t,
+    })
```
Comment (minotogna) [inline src/server/controller/cycleData/updateOriginalDataPoint.ts]: there should be only one method invoked

File: src/server/controller/cycleData/updateOriginalDataPoint.ts [hunk 4]
```diff
@@ -1,9 +1,178 @@
...
+type updateDependantsProps = {
+  countryIso: CountryIso
+  assessment: Assessment
+  cycle: Cycle
+  user: User
+  colName: string
+  updatedOriginalDataPoint: OriginalDataPoint
+  client: BaseProtocol
+}
+
```
Comment (minotogna) [inline src/server/controller/cycleData/updateOriginalDataPoint.ts]: this is unreadable. extract methods

File: src/client/store/pages/assessmentSection/slice.ts (hunk 1 above)
Comment (minotogna) [reply src/client/store/pages/assessmentSection/slice.ts]: this should not be changed. backend should pass something like:
["FIN-fra-2020-originalDataPointValue-otherWoodedLand-2020-nodeUpdate",[{"value":{"raw":"65346345.00","estimated":true,"validation":{"valid":false,"messages":[{"key":"extentOfForest.fedAreasExceedTotalLandArea"}]}}}]]

File: src/server/controller/cycleData/updateOriginalDataPoint.ts (hunk 4 above)
Comment (sorja) [reply src/server/controller/cycleData/updateOriginalDataPoint.ts]: yes

File: src/server/controller/cycleData/updateOriginalDataPoint.ts (hunk 3 above)
Comment (sorja) [reply src/server/controller/cycleData/updateOriginalDataPoint.ts]: yes

File: src/client/store/pages/assessmentSection/slice.ts (hunk 1 above)
Comment (sorja) [reply src/client/store/pages/assessmentSection/slice.ts]: yes

File: src/client/pages/AssessmentSection/DataTable/Table/Row/RowData/Cell/Cell.tsx [hunk 5]
```diff
@@ -66,7 +66,16 @@ const Cell: React.FC<Props> = (props) => {
...
   const className = useClassName({ col, row, tableName, valid })
   const { onChange, onChangeNodeValue, onPaste } = useOnChange({ table, col, row, nodeValue, data, sectionName })
-  useListenNodeUpdate({ countryIso, assessmentName, cycleName, tableName, variableName, colName })
+
+  // If cell is ODP, listen to odp changes
+  useListenNodeUpdate({
+    countryIso,
+    assessmentName,
+    cycleName,
+    tableName: nodeValue?.odp ? 'originalDataPointValue' : tableName,
```
Comment (minotogna) [inline src/client/pages/AssessmentSection/DataTable/Table/Row/RowData/Cell/Cell.tsx]: don't we have a contact for 'originalDataPointValue'?

File: src/client/store/pages/assessmentSection/hooks/index.ts [hunk 6]
```diff
@@ -15,9 +15,7 @@ export const useTableSections = (props: { sectionName: string }): Array<TableSec
 
 const useOriginalDataPointData = (): Record<string, Record<string, NodeValue>> | undefined => {
   const countryIso = useCountryIso()
-  return useAppSelector(
-    (state) => state.pages.assessmentSection.originalDataPointData?.[countryIso]?.originalDataPointValue
-  )
+  return useAppSelector((state) => state.pages.assessmentSection.data?.[countryIso]?.originalDataPointValue)
```
Comment (minotogna) [inline src/client/store/pages/assessmentSection/hooks/index.ts]: I would use a constant here as well

File: src/server/controller/cycleData/updateOriginalDataPoint/getForestCharacteristicsDependantUpdates.ts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    otherPlantedForestArea: String(
+      ODPs.calcTotalSubFieldArea({
+        originalDataPoint: updatedOriginalDataPoint,
+        field: 'forestPercent',
+        subField: 'otherPlantedForestPercent',
+      })
+    ),
+  }
+
+  const tableName = 'originalDataPointValue'
```
Comment (minotogna) [inline src/server/controller/cycleData/updateOriginalDataPoint/getForestCharacteristicsDependantUpdates.ts]: use a constant

File: src/server/controller/cycleData/updateOriginalDataPoint/index.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    const forestCharacteristicsUpdates = await getForestCharacteristicsDependantUpdates({
+      countryIso,
+      assessment,
+      cycle,
+      colName,
+      user,
+      updatedOriginalDataPoint,
+      client: t,
+    })
+
```
Comment (minotogna) [inline src/server/controller/cycleData/updateOriginalDataPoint/index.ts]: can't we call only one method ?

File: src/server/controller/cycleData/updateOriginalDataPoint/updateDependantsProps.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+import { BaseProtocol } from '@server/db'
+
+export type updateDependantsProps = {
+  countryIso: CountryIso
+  assessment: Assessment
+  cycle: Cycle
+  user: User
+  colName: string
+  updatedOriginalDataPoint: OriginalDataPoint
```
Comment (minotogna) [inline src/server/controller/cycleData/updateOriginalDataPoint/updateDependantsProps.ts]: can't this be called originalDataPoint directly ? also, use upper case for types

File: src/server/controller/cycleData/updateOriginalDataPoint/index.ts [hunk 10]
```diff
@@ -16,11 +17,13 @@ export const updateOriginalDataPoint = async (
...
+    { assessment, cycle, originalDataPoint },
+    client
+  )
+
   return client.tx(async (t) => {
-    const updatedOriginalDataPoint = await OriginalDataPointRepository.update(
-      { assessment, cycle, originalDataPoint },
-      t
-    )
+    await updateDependantNodes({ originalDataPoint: updatedOriginalDataPoint, assessment, cycle, user }, t)
```
Comment (minotogna) [inline src/server/controller/cycleData/updateOriginalDataPoint/index.ts:26]: cant we just reuse the logic from src/server/controller/cycleData/persistNodeValue/index.ts line 18-31? why reimplementing the logic here?

File: src/test/integration/assessment/originalDataPoint.ts [hunk 11]
```diff
@@ -47,7 +50,11 @@ export default (): void =>
 
     it('Edit existing/not existing Original data point', async () => {
       const editedOriginalDataPoint = await CycleDataController.updateOriginalDataPoint({
-        assessment,
+        assessment: {
+          ...assessment,
+          // Test assessment meta cache is empty
+          metaCache: { ...assessment.metaCache, calculations: { dependants: {}, dependencies: {} } },
+        },
```
Comment (minotogna) [inline src/test/integration/assessment/originalDataPoint.ts]: cant this be added to the test assessment directly ?

File: src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts (hunk 2 above)
Comment (minotogna) [inline src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts:11]: check if value is not needed in src/server/controller/cycleData/persistNodeValue/index.ts

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/meta/assessment/table.ts [hunk 12]
```diff
@@ -14,6 +14,7 @@ export enum TableNames {
   primaryDesignatedManagementObjective = 'primaryDesignatedManagementObjective',
 
   valueAggregate = 'value_aggregate',
+  originalDataPointValue = 'originalDataPointValue',
```
Comment (minotogna) [inline src/meta/assessment/table.ts]: cant we have alphabetical order?

File: src/client/store/pages/assessmentSection/slice.ts [hunk 13]
```diff
@@ -60,7 +59,11 @@ export const assessmentSectionSlice = createSlice({
     })
 
     builder.addCase(getOriginalDataPointData.fulfilled, (state, { payload }) => {
-      state.originalDataPointData = payload
+      const countryIso = Object.keys(payload || {})[0] as CountryIso
+      if (countryIso) {
```
Comment (minotogna) [inline src/client/store/pages/assessmentSection/slice.ts]: why shouldn't we have countryIso here?

File: src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts (hunk 2 above)
Comment (sorja) [reply src/server/controller/cycleData/persistNodeValue/calculateDependantNodes/calculateDependantNodes.ts:11]: It is not used
https://github.com/openforis/fra-platform/blob/e2031fc11b7e832653914f6545f59a1081fce7ff/src/server/controller/cycleData/persistNodeValue/index.ts#L18

File: src/test/integration/assessment/originalDataPoint.ts (hunk 11 above)
Comment (sorja) [reply src/test/integration/assessment/originalDataPoint.ts:57]: assessment fetched from db, should we put it in db..?

https://github.com/openforis/fra-platform/blob/e2031fc11b7e832653914f6545f59a1081fce7ff/src/test/integration/assessment/originalDataPoint.ts#L21

File: src/server/controller/cycleData/updateOriginalDataPoint/index.ts (hunk 10 above)
Comment (minotogna) [reply src/server/controller/cycleData/updateOriginalDataPoint/index.ts:26]: we should talk in person

File: src/test/integration/assessment/originalDataPoint.ts [hunk 14]
```diff
@@ -47,7 +50,9 @@ export default (): void =>
 
     it('Edit existing/not existing Original data point', async () => {
       const editedOriginalDataPoint = await CycleDataController.updateOriginalDataPoint({
-        assessment,
+        assessment: {
+          ...assessment,
+        },
```
Comment (minotogna) [inline src/test/integration/assessment/originalDataPoint.ts:55]: why just spreading itself ? is it needed ?

File: src/server/repository/assessment/assessment/createAssessment.ts [hunk 15]
```diff
@@ -7,15 +7,17 @@ import { read } from '@server/repository/assessment/assessment/read'
 
 export const createAssessment = async (
   params: {
-    assessment: Pick<Assessment, 'props'>
+    assessment: Pick<Assessment, 'props' | 'metaCache'>
```
Comment (minotogna) [inline src/server/repository/assessment/assessment/createAssessment.ts:10]: when creating an assessment, metaCache is not populated. 
can u avoid passing metaCache here?

File: src/test/integration/assessment/originalDataPoint.ts (hunk 14 above)
Comment (sorja) [reply src/test/integration/assessment/originalDataPoint.ts:55]: typo

File: src/server/repository/assessment/assessment/createAssessment.ts (hunk 15 above)
Comment (sorja) [reply src/server/repository/assessment/assessment/createAssessment.ts:10]: how should we populate the test assessment?
we fetch it from db and if we dont add it there, dont add it here, then.. what is left?

Comment (minotogna) [reply src/server/repository/assessment/assessment/createAssessment.ts:10]: check my pr

**Pull request closed without merging**